src/server/engine_shard_set.cc

```diff
@@ -525,23 +525,14 @@ void EngineShard::PollExecution(const char* context, Transaction* trans) {
       continuation_trans_ = head;
   }
 
-  // Either the poll had no caller or it was handled above
-  if (trans == nullptr)
-    return;
-
-  // If the pointer is valid, we didn't handle it above, so trans is still armed
-  DCHECK(trans->IsArmedInShard(sid));
-
-  // OOO means no transaction before us in the txq accesses our keys, so we can run earlier
-  bool is_ooo = local_mask & Transaction::OUT_OF_ORDER;
-
-  // Still suspended shards need to run just to finalize and unregister, so we can run anytime
-  bool is_suspended = local_mask & Transaction::SUSPENDED_Q;
-  DCHECK_EQ(local_mask & Transaction::AWAKED_Q, 0);
-
-  if (is_ooo || is_suspended) {
+  // If trans was already handled, the pointer was cleared. Otherwise we can run it now if:
+  // 1. It is OUT_OF_ORDER, so no one before accesses it's keys
+  // 2. it is SUSPENDED_Q, so it has only to clean up stale shards after waking up
+  if (trans && (local_mask & (Transaction::OUT_OF_ORDER | Transaction::SUSPENDED_Q)) > 0) {
     DCHECK(trans != head);
+    CHECK(trans->DisarmInShard(sid));
```

Comment:
Don't you already disarm it  at line 451?

Reply:
We fetch the local mask, but we disarm only if AWAKED was present. See the comment of `DisarmWhen`.

The other option is disarming if OUT_OF_ORDER or AWAKED are present and then checking `head == trans || head->Disarm` which is also no that elegant

The third option is just reading LocalMask and Disarming seprately, like we did before, but it's not that elegant and grows the outside interface, where we only read local mask to disarm them

Reply:
oh, I saw the semantics of DisarmInShardWhen and I find it a bit confusing. You returned a mask if we are armed but you disarm only if the mask matches the argument. And now we need to track somehow whether we disarmed the state or not based on the mask returned.  I do not have a good suggestion, just stating it was confusing to follow. maybe we should add some comments to line 451.

Reply:
> And now we need to track somehow whether we disarmed the state or not based on the mask returned

Well, we enter the if block if we disarmed. Otherwise we assume we didn't. Will add comments